server/datastore/mysql/software.go

```diff
@@ -71,31 +61,223 @@ func uniqueStringToSoftware(s string) fleet.Software {
+	reported map[string]struct{},
+	stored []fleet.HostSoftwareInstalledPath,
+	hostSoftware []fleet.Software,
+) (
+	toInsert []fleet.HostSoftwareInstalledPath,
+	toDelete []uint,
+	err error,
+) {
+	if len(reported) != 0 && len(hostSoftware) == 0 {
+		// Error condition, something reported implies that the host has some software
+		err = fmt.Errorf("software installed paths for host %d were reported but host contains no software", hostID)
+		return
+	}
+
+	sIDLookup := map[uint]fleet.Software{}
+	for _, s := range hostSoftware {
+		sIDLookup[s.ID] = s
+	}
+
+	sUnqStrLook := map[string]fleet.Software{}
+	for _, s := range hostSoftware {
+		sUnqStrLook[s.ToUniqueStr()] = s
+	}
+
+	iSPathLookup := make(map[string]fleet.HostSoftwareInstalledPath)
+	for _, r := range stored {
+		s, ok := sIDLookup[r.SoftwareID]
+		// Software currently not found on the host, should be deleted ...
+		if !ok {
+			toDelete = append(toDelete, r.ID)
+			continue
+		}
+
+		key := fmt.Sprintf("%s%s%s", r.InstalledPath, fleet.SoftwareFieldSeparator, s.ToUniqueStr())
+		iSPathLookup[key] = r
+
+		// Anything stored but not reported should be deleted
+		if _, ok := reported[key]; !ok {
+			toDelete = append(toDelete, r.ID)
+		}
+	}
+
+	for key := range reported {
+		parts := strings.SplitN(key, fleet.SoftwareFieldSeparator, 2)
+		iSPath, unqStr := parts[0], parts[1]
+
+		// Shouldn't be possible ... everything 'reported' should be in the the software table
+		// because this executes after 'ds.UpdateHostSoftware'
+		s, ok := sUnqStrLook[unqStr]
+		if !ok {
+			err = fmt.Errorf("reported installed path for %s does not belong to any stored software entry", unqStr)
+			return
+		}
+
+		if _, ok := iSPathLookup[key]; ok {
+			// Nothing to do
+			continue
+		}
+
+		toInsert = append(toInsert, fleet.HostSoftwareInstalledPath{
+			HostID:        hostID,
+			SoftwareID:    s.ID,
+			InstalledPath: iSPath,
+		})
+	}
+
+	return
+}
+
+func deleteHostSoftwareInstalledPaths(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	toDelete []uint,
+) error {
+	if len(toDelete) == 0 {
+		return nil
+	}
+
+	stmt := `DELETE FROM host_software_installed_paths WHERE id IN (?)`
+	stmt, args, err := sqlx.In(stmt, toDelete)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "building delete statement for delete host_software_installed_paths")
+	}
+	if _, err := tx.ExecContext(ctx, stmt, args...); err != nil {
+		return ctxerr.Wrap(ctx, err, "executing delete statement for delete host_software_installed_paths")
+	}
+
+	return nil
+}
+
+func insertHostSoftwareInstalledPaths(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	toInsert []fleet.HostSoftwareInstalledPath,
+) error {
+	if len(toInsert) == 0 {
+		return nil
+	}
+
+	stmt := "INSERT INTO host_software_installed_paths (host_id, software_id, installed_path) VALUES %s"
+	batchSize := 500
+
+	for i := 0; i < len(toInsert); i += batchSize {
+		end := i + batchSize
+		if end > len(toInsert) {
+			end = len(toInsert)
+		}
+		batch := toInsert[i:end]
```

Comment:
Are you not inserting 501 items instead of 500 🤔 ?

Reply:
I'll double check, but I don't think so. When specifying two indices the interval is half open, meaning it includes the lower bound but not the upper bound, but I might be wrong. Thanks for bringing it up!